pkg/services/authn/authnimpl/sync/rbac_sync.go

```diff
@@ -77,7 +80,8 @@ func (s *RBACSync) fetchPermissions(ctx context.Context, ident *authn.Identity)
 
 	permissions := make([]accesscontrol.Permission, 0, 8)
 	roles := ident.ClientParams.FetchPermissionsParams.Roles
-	if len(roles) > 0 {
+	actions := ident.ClientParams.FetchPermissionsParams.AllowedActions
+	if len(roles) > 0 || len(actions) > 0 {
 		for _, role := range roles {
 			roleDTO, err := s.ac.GetRoleByName(ctx, ident.GetOrgID(), role)
 			if err != nil && !errors.Is(err, accesscontrol.ErrRoleNotFound) {
```

Comment:
If we use the permissions key to store both roles and actions, you can't return an error on L89 the way its currently doing.

Reply:
Also, you may wanna deduplicate the permissions built from roles+permissions in `permissions`. As their likely would be duplicates until we are sure we don't need roles anymore.

Reply:
> This might also be why I am seeing a 401 on main when trying the heterogeneous permissions value in my testing.

I think if you get an error in that scenario, it would more likely be a 403 Forbidden. This usually happens if we're unable to fetch roles from the database, preventing proper authorization.
You getting 401s is still problematic though and probably points out to invalid credentials :sweat_smile: To help pinpoint the exact cause, could you share any relevant logs or traces?
Btw, the separation of roles and actions within permissions should already be handled by the ext_jwt client at this point.

> Also, you may wanna deduplicate the permissions built from roles+permissions in permissions. As their likely would be duplicates until we are sure we don't need roles anymore.

I agree, but considering the expected number of permissions within a token, I don't anticipate significant performance issues from potential duplicates. We can prioritize de-duplication later if needed. One approach could be revisiting the `GroupScopesByActions` function, which up until recently handled this.